### ISSUE TITLE
Cancelling an 'iput -f' breaks the integrity of a replica whilst preserving the previous checksum.

### ISSUE THREAD

**Issue by triitaojcsc** [irods/irods#3091]
- [x] master
- [x] 4-2-stable
-------

Cancelling an `iput -f` breaks the integrity of a replica whilst preserving the previous checksum.
# Environment

```
[user@localhost ~]$ cat /etc/redhat-release
CentOS release 6.6 (Final)

[user@localhost ~]$ rpm -qa | grep irods
irods-database-plugin-postgres93-1.6-0.x86_64
irods-icat-4.1.5-0.x86_64
```

Note: Also reproducible on 4.1.8.

```
[user@localhost ~]$ ilsresc --tree
testrepl:replication
├── res3
└── res4
```
# Changes in core.re

```
acPostProcForPut {
    msiSysChksumDataObj;
}
acPostProcForRepl {
    msiSysChksumDataObj;
}
```
# Starting situation

User has uploaded a file '100M', which has two replicas, two identical checksums and two identical file sizes.

```
[user@localhost ~]$ ils -L 100M
  user          0 testrepl;res4    104857600 2016-03-09.11:20 & 100M
    sha2:IEkqTQ2E+L6xdn9mFiKfhdRMKCe2S9v7Jg7hL6EQng4=    generic    /var/lib/irods/iRODS/Vault-res4/home/user/100M
  user          1 testrepl;res3    104857600 2016-03-09.11:20 & 100M
    sha2:IEkqTQ2E+L6xdn9mFiKfhdRMKCe2S9v7Jg7hL6EQng4=    generic    /var/lib/irods/iRODS/Vault-res3/home/user/100M
```
# Action that creates the problem

User decides to overwrite the '100M' file with a 2147483648 byte (2GB) file, but cancels the transfer.

```
[user@localhost ~]$ iput -f 2GB 100M
^C
```

(Cancelling can be done with CTRL-C, but note: for testing purposes perhaps timeouts should also be simulated? Pretty sure this also occurs when the connection is abrubtly closed for whatever reasons.)
# The problem

Now the user sees two replicas, two identical checksums and two differing file sizes. We can see that the 2147483648 byte file did not transfer completely, because the user cancelled the transfer.

```
[user@localhost ~]$ ils -L 100M
  user          0 testrepl;res4   2020194304 2016-03-09.11:20 & 100M
    sha2:IEkqTQ2E+L6xdn9mFiKfhdRMKCe2S9v7Jg7hL6EQng4=    generic    /var/lib/irods/iRODS/Vault-res4/home/user/100M
  user          1 testrepl;res3    104857600 2016-03-09.11:20 & 100M
    sha2:IEkqTQ2E+L6xdn9mFiKfhdRMKCe2S9v7Jg7hL6EQng4=    generic    /var/lib/irods/iRODS/Vault-res3/home/user/100M
```

Two differing replicas that have identical checksums but differing sizes seems like a bad idea. I'd assume that it's very common for iRODS admins to survey the integrity of the files by comparing the checksums instead of file sizes.
# Desired outcome

I'm not sure what the solution is. How would you solve this?

I don't like the idea that an unfinished transfer breaks the integrity of the other replica, but leaves the checksum untouched. I think `iput -f` file transfer cancellations should be handled very robustly. And if something goes wrong, it should be obvious from the checksums in the ICAT database.
# Additional bonus

The data_is_dirty bit is '1' for both replicas. Meaning that the replicas are considered clean (because of the convoluted way that somebody has decided that '1' is 'no' and '0' is 'yes').

```
ICAT=> select data_repl_num,data_size,data_is_dirty,data_checksum,create_ts,modify_ts,resc_hier from r_data_main where data_name = '100M';
 data_repl_num | data_size  | data_is_dirty |                   data_checksum                   |  create_ts  |  modify_ts  |   resc_hier
---------------+------------+---------------+---------------------------------------------------+-------------+-------------+---------------
             1 |  104857600 |             1 | sha2:IEkqTQ2E+L6xdn9mFiKfhdRMKCe2S9v7Jg7hL6EQng4= | 01457515240 | 01457515240 | testrepl;res3
             0 | 2020194304 |             1 | sha2:IEkqTQ2E+L6xdn9mFiKfhdRMKCe2S9v7Jg7hL6EQng4= | 01457515239 | 01457515239 | testrepl;res4
(2 rows)
```

-- Taneli Riitaoja / CSC - IT Center for Science Ltd.

**Comment by trel** [comment]
This is a fantastic bug report - thank you.

Upon initial review, I'd think the solution is to mark the replica 'stale/dirty' that received the updated file size (but did not complete the transfer successfully).  This should allow any subsequent connections/actions to take appropriate action.

The `data_is_dirty` history hinges on the definition of dirty.  It was originally used in the context of 'in flight', so a dirty copy was one that was 'active' or 'new/newest'.  This is confusing because we're nearly always more interested in the data at rest, where 'dirty', or 'stale', means its integrity is in question or known to be bad.

**Comment by mcast** [comment]
Wow!

> I'm not sure what the solution is. How would you solve this?

As a user, I would at least like the checksum on the `-n1` (partially replaced) copy to be cleared before it can be seen on the new data - either before the data on disk is lost, or immediately after.  Assuming it is done by a clobbering open, then before (including the iCat `COMMIT`) seems safest.

At the point of discovering the partially `iput` file, I would have to wonder which is the valid file and how I get back to a sane state.  This presumably depends on why some process was writing the new data.

**Comment by triitaojcsc** [comment]
> Upon initial review, I'd think the solution is to mark the replica 'stale/dirty' that received the updated file size (but did not complete the transfer successfully). This should allow any subsequent connections/actions to take appropriate action.

Hmm, the dirty bit handling is pretty complex. I will maybe make another ticket in the future how 'rebalance' gets confused in some cases when the ICAT database for some reason has both replicas' dirty bits as '0'. When rebalance encounters a file like this, it just throws its hands up in the air. Which means that I don't have any useful tools to balance my replication trees. Not that rebalance is an useful too, it lacks options and verbose mode ;) But I digress.

It occurs to me that if we lived in a beautiful world, an iput -f file transfer might work like this:

1. File exist in iRODS and has two replicas (in a replication tree), for example /disk/file.dat and /tape/file.dat.
2. An iput -f operation for file.dat is issued. File transfer starts, but instead of just directly starting to overwrite /disk/file.dat ...
3. ... a file /disk/file.dat.<identifier> is created instead and the data is transferred to it.
4. When the file transfer finishes, two things (A+B) happen in an atomic operation:
    A /disk/file.dat is overwritten by /disk/file.dat.<identifier>
    B Whatever hardcoded information that is required gets written to the ICAT database
5. Normal operation resumes (rule files, replication, etc.)

But that's probably a bigger change (by an order of magnitude) to implement :p 

-- Taneli

**Comment by jasoncoposky** [comment]
We are working on a new transport protocol creatively called 'chunky transfer' (starting here https://github.com/DICE-UNC/chunky) as a replacement for the existing restart functionality that will address a lot of this.

One of the first things we did was update the file size in the catalog even on a failure in order to keep things in sync as much as possible.  We should include the checksum in the update were one to exist previously and consider what to do with the dirty flag.

**Comment by trel** [comment]
When the file size is updated in the catalog, but errors out, the checksum should be cleared in rsDataObjClose()

**Comment by trel** [comment]
proposal:
- data_is_dirty, and checksum are cleared in rsDataObjOpen()
- size is updated in rsDataObjClose() (already happening)
- on success... rsDataObjClose() would update the data_is_dirty, and checksum
- on failure... rsDataObjClose() would not touch them, they would remain 'stale' and 'no checksum'.

**Comment by bh9** [comment]
Does iput act on replicas sequentially or in parallel? If sequentially, then this method will only ever mark 1 replica as dirty (and without checksum) so that would work. If parallel, then this would leave no "good" replicas which is a state we would like to avoid

**Comment by trel** [comment]
Just to follow up and confirm your edit...

An `iput -f` would only overwrite a single replica.  If others were to be updated, due to rules or policy baked into a replication coordinating resource, those would be handled serially after the initial overwrite was complete.

So, agreed, this should be okay and never leave the system in a state where there would be no good replicas (unless there was only a single original replica to begin with, which seems defensible/expected).

**Comment by alanking** [comment]
Related to #5496

**Comment by alanking** [comment]
This appears fixed in 4.2.9:
```
$ ils -L big
  rods              0 repl1;child2     41943040 2021-04-29.16:50 & big
    sha2:gKNyEYjkAhiwiyZ3a8U72ugeR4T/9x1xRQoZcxnLoRM=    generic    /var/lib/irods/child2vault/home/rods/big
  rods              1 repl1;child1     41943040 2021-04-29.16:50 & big
    sha2:gKNyEYjkAhiwiyZ3a8U72ugeR4T/9x1xRQoZcxnLoRM=    generic    /var/lib/irods/child1vault/home/rods/big
$ iput -K -f -R repl1 bigger big 
^C
$ ils -L
/tempZone/home/rods:
  rods              0 repl1;child2     83886080 2021-04-29.16:51 X big
    sha2:M6OhHVTejt5gTCQ87f3h70tTTV6jJ5yd1X3zFARcI98=    generic    /var/lib/irods/child2vault/home/rods/big
  rods              1 repl1;child1     41943040 2021-04-29.16:50 & big
    sha2:gKNyEYjkAhiwiyZ3a8U72ugeR4T/9x1xRQoZcxnLoRM=    generic    /var/lib/irods/child1vault/home/rods/big
```
Should I add a test and close?

**Comment by trel** [comment]
yes please.   this issue gets a 5y gold star.

**Comment by trel** [comment]
checkboxed.
